src/policy/ephemeral_policy.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2024-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <policy/ephemeral_policy.h>
+#include <policy/policy.h>
+
+bool HasDust(const CTransactionRef& tx, CFeeRate dust_relay_rate)
+{
+    return std::any_of(tx->vout.cbegin(), tx->vout.cend(), [&](const auto& output) { return IsDust(output, dust_relay_rate); });
+}
+
+bool CheckValidEphemeralTx(const CTransactionRef& tx, CFeeRate dust_relay_rate, CAmount base_fee, CAmount mod_fee, TxValidationState& state)
+{
+    // We never want to give incentives to mine this transaction alone
+    if ((base_fee != 0 || mod_fee != 0) && HasDust(tx, dust_relay_rate)) {
+        return state.Invalid(TxValidationResult::TX_NOT_STANDARD, "dust", "tx with dust output must be 0-fee");
+    }
+
+    return true;
+}
+
+std::optional<Txid> CheckEphemeralSpends(const Package& package, CFeeRate dust_relay_rate, const CTxMemPool& tx_pool)
```

Comment:
Skimming this, two thoughts come to mind:

 * why call it "check ephemeral spends" when "check dust is swept" is right there?
 * I'm surprised this check isn't the other way around, something like:

```c++
std::set<COutPoint> dust_created;
for (tx : package) {
     for (tx_input : tx->vin) {
        dust_created.erase(tx_input.prevout);
     }
     Txid txid = tx->GetHash();
     for (int out = 0; out < tx->vout.size(); ++out) {
          const auto* tx_output = tx->vout[out];
          if (IsDust(tx_output)) {
             dust_created.insert(COutPoint{txid, out});
          }
    }
}
if (!dust_created.empty()) {
    ...
}
```

If the idea is that all dust created by a tx should be swept by the immediate next tx (so that the package can't be split and leave dust in the utxo set), you could move the empty() check to the middle of the main loop. Reporting the txid of the transaction that created the dust seems more logical than the one that should have spent it but didn't to me, fwiw.

Reply:
> why call it "check ephemeral spends" when "check dust is swept" is right there?

Er, dunno, I chose the name 2 years ago I think. Could amend the name change I have in the follow-up!

> I'm surprised this check isn't the other way around, something like:

There are two interlocking rules:
1) If a transaction has dust, it must be 0-fee (CheckValidEphemeralTx)
2) If a transaction is spending from a transaction that has unconfirmed dust, the dust must be spent (CheckEphemeralSpends)

The snippet above is different in at least a couple of ways, the more important being that it would not detect the second case where there are in-mempool ancestors to the package.

> If the idea is that all dust created by a tx should be swept by the immediate next tx 

p2p relay questions aside, it's not really the case. Batched CPFP is possible for one, and if/when `submitpackage` is generalized further, could allow arbitrary chains of transactions with single dust outputs.

Reply:
> The snippet above is different in at least a couple of ways, the more important being that it would not detect the second case where there are in-mempool ancestors to the package.

Sure, I was handwaving that any in-mempool ancestors were already part of the package.

Mostly, I think I was misled by:

```
 * All the dust is spent if TxA+TxB+TxC is accepted, but the mining template may just pick
 * up TxA+TxB rather than the three "legal configurations:
 * 1) None
 * 2) TxA+TxB+TxC
 * 3) TxA+TxC
```

(missing close quote after `"legal` btw)

But we don't consider (2) legal, and can't because our current mining algorithm isn't smart enough to be able to call/satisfy `CheckEphemeralSpends()`. Cluster mempool could conceivably be smart enough to preserve that property, though, by ensuring that txs are never split from a chunk if it would cause a function like this to be unhappy... (In that case, the chunk would likely be [TxA, TxC, TxB] anyway...) In any event, relaying TxB prior to TxA being confirmed probably increases bad incentives for leaving dust in the utxo set anyway...

(This rationale would be better in a BIP than in code comments... /grumble)

Reply:
> But we don't consider (2) legal

Right this could have been phrased better, "legal" here means a block template may be created where there is no dust entering into the utxo set. ie If TxA is mined, we want TxC mined, not that the implementation would allow that config to happen currently. 

>  Cluster mempool could conceivably be smart enough to preserve that property, though, by ensuring that txs are never split from a chunk if it would cause a function like this to be unhappy... 

I noodled on stuff like this for a while half a year ago and came away unsatisfied. I think it's hard to get something more advanced that acts natural and non-pinny from a wallet perspective but I don't have notes in front of me. 

> (This rationale would be better in a BIP than in code comments... /grumble)

Enjoy the aborted BIP attempt https://github.com/bitcoin/bips/pull/1524